staging/src/k8s.io/client-go/tools/cache/shared_informer.go

```diff
@@ -163,6 +163,18 @@ type SharedInformer interface {
 	// store. The value returned is not synchronized with access to the underlying store and is not
 	// thread-safe.
 	LastSyncResourceVersion() string
+
+	// The WatchErrorHandler is called whenever ListAndWatch drops the
+	// connection with an error. After calling this handler, the informer
+	// will backoff and retry.
+	//
+	// The default implementation looks at the error type and tries to log
+	// the error message at an appropriate level.
+	//
+	// Implementations of this handler may display the error message in other
+	// ways. Implementations should return quickly - any expensive processing
+	// should be offloaded.
+	SetWatchErrorHandler(handler WatchErrorHandler) error
```

Comment:
Shared informers are usually shared :). This interface should attempt to reflect that ownership model by exposing functionality that is safe to consume at multiple sites across potentially large applications. SharedInformers differ a lot from regular informers to make sure that's the case (see cache mutation detection, and shared processor). This function stands out to me because it is:
* Exclusive: last call before starting the informer wins. Previous calls are silently ignored.
* Has potential for unknown/unbounded/hard to notice spillover effects on other users of the same shared informer.

Reply:
I don't think this is the mechanism we'll use to get controllers to e.g. stop operating during a disconnection. This is just for visibility.

I think we can state these things in the comment and that will be sufficient.

Reply:
Sorry for not being clear, I think we should add to the comment here:

* There's only one handler, so if you call this multiple times, last one wins; calling after the informer has been started returns an error.
* The handler is intended for visibility, not to e.g. pause the consumers.

(these points don't need to go in the place where you copied this comment, just here--since they're unique to shared informers, or at least the first one is.)

Reply:
sorry - github aggressively hides conversations marked 'resolved' so i missed that this was still outstanding! I like your suggestions, added!